src/tree/hist/histogram.h

```diff
@@ -50,50 +249,159 @@ template <typename GradientSumT, typename ExpandEntry> class HistogramBuilder {
     buffer_.Init(total_bins);
     builder_ = common::GHistBuilder<GradientSumT>(n_threads, total_bins);
     is_distributed_ = is_distributed;
+    max_depth_ = std::max(max_depth, 1);
+    colsample_bytree_ = colsample_bytree;
+    colsample_bylevel_ = colsample_bylevel;
+    colsample_bynode_ = colsample_bynode;
+    column_sampler_ = column_sampler;
+    if (histograms_buffer.size() == 0) {
+      n_bins_ = total_bins;
+      histograms_buffer.resize(n_threads);
+      local_threads_mapping.resize(n_threads);
+      offsets64_.resize(n_threads);
+      #pragma omp parallel num_threads(n_threads)
+      {
+        const size_t tid = omp_get_thread_num();
+        local_threads_mapping[tid].resize((1 << (max_depth_ + 2)));
+        histograms_buffer[tid].resize((1 << (max_depth_ - 1)));
+        offsets64_[tid].resize((1 << (max_depth_ - 1)));
+      }
+    }
   }
 
-  template <bool any_missing>
-  void
-  BuildLocalHistograms(DMatrix *p_fmat,
-                       std::vector<ExpandEntry> nodes_for_explicit_hist_build,
-                       common::RowSetCollection const &row_set_collection,
-                       const std::vector<GradientPair> &gpair_h) {
-    const size_t n_nodes = nodes_for_explicit_hist_build.size();
-
-    // create space of size (# rows in each node)
-    common::BlockedSpace2d space(
-        n_nodes,
-        [&](size_t node) {
-          const int32_t nid = nodes_for_explicit_hist_build[node].nid;
-          return row_set_collection[nid].Size();
-        },
-        256);
-
-    std::vector<GHistRowT> target_hists(n_nodes);
-    for (size_t i = 0; i < n_nodes; ++i) {
-      const int32_t nid = nodes_for_explicit_hist_build[i].nid;
-      target_hists[i] = hist_[nid];
+  template<typename BinIdxType, bool any_missing, bool hist_fit_to_l2, typename BuildHistFunc>
+  BuildHistFunc GetBuildHistSrategy(int depth, const bool any_sparse_column) {
+    // now column sampling supported only for missings due to fid_least_bins_ set
+    if (any_missing
+        && (colsample_bytree_ < 0.1 || colsample_bylevel_ < 0.1)
```

Comment:
These are lots of specializations ....

Reply:
But fortunately there is no need to have any explicit specialization.
Would it be better to use general bool constants instead of template arguments? It would lead to more if conditions to select specific optimized kernel.